### ISSUE TITLE
[rust binding] broken immutable invariant on `QueryMatch.captures`

### ISSUE THREAD

**Issue by kawaemon** [tree-sitter/tree-sitter#3537]
### Problem

`tree_sitter::QueryMatch.captures` field is immutable slice of `QueryCapture`, therefore it must be immutable while in scope.

However calling `<QueryMatch as Iterator>::next` may overwrite `captures` field of previously produced `QueryCapture`, resulting causing unsoundness.

### Steps to reproduce

This code:
```rs
fn main() {
    let src = r#"
        public class Test {
            private String a;
            private int b;
            private Integer c;
        }
    "#
    .as_bytes();

    let lang = &tree_sitter_java::language();
    let mut parser = Parser::new();
    parser.set_language(lang).unwrap();
    let tree = parser.parse(src, None).unwrap();

    let query = Query::new(lang, "(field_declaration) @decl").unwrap();
    let mut cursor = QueryCursor::new();

    let mut matches = cursor.matches(&query, tree.root_node(), src);

    let match1 = matches.next().unwrap();
    println!("m1: {:?}", match1.captures);
    println!();

    let match2 = matches.next().unwrap();
    println!("m1: {:?}", match1.captures);
    println!("m2: {:?}", match2.captures);
    println!();

    let match3 = matches.next().unwrap();
    println!("m1: {:?}", match1.captures);
    println!("m2: {:?}", match2.captures);
    println!("m3: {:?}", match3.captures);
}
```

outputs below. Notice that `m1` and `m2` are mutating everytime calling `matches.next()`.
```
m1: [QueryCapture { node: {Node field_declaration (2, 12) - (2, 29)}, index: 0 }]

m1: [QueryCapture { node: {Node field_declaration (3, 12) - (3, 26)}, index: 0 }]
m2: [QueryCapture { node: {Node field_declaration (3, 12) - (3, 26)}, index: 0 }]

m1: [QueryCapture { node: {Node field_declaration (4, 12) - (4, 30)}, index: 0 }]
m2: [QueryCapture { node: {Node field_declaration (4, 12) - (4, 30)}, index: 0 }]
m3: [QueryCapture { node: {Node field_declaration (4, 12) - (4, 30)}, index: 0 }]
```

### Expected behavior

should output
```java
m1: [QueryCapture { node: {Node field_declaration (2, 12) - (2, 29)}, index: 0 }]

m1: [QueryCapture { node: {Node field_declaration (2, 12) - (2, 29)}, index: 0 }]
m2: [QueryCapture { node: {Node field_declaration (3, 12) - (3, 26)}, index: 0 }]

m1: [QueryCapture { node: {Node field_declaration (2, 12) - (2, 29)}, index: 0 }]
m2: [QueryCapture { node: {Node field_declaration (3, 12) - (3, 26)}, index: 0 }]
m3: [QueryCapture { node: {Node field_declaration (4, 12) - (4, 30)}, index: 0 }]
```

### Tree-sitter version (tree-sitter --version)

```
[dependencies]
tree-sitter = "0.22.6"
tree-sitter-java = "0.21.0"
```

### Operating system/version

ArchLinux

**Comment by WillLillis** [comment]
Looks like a duplicate of #2843 (also see #2847 for a more concise example).

Potentially fixed by #3504

**Comment by kawaemon** [comment]
ah looks like so.
Closing as duplicated issue. Thank you!
